### ISSUE TITLE
Screen flickering when 1000th line enters the screen combined with horizontal scrollbar

### ISSUE THREAD

**Issue by thg2k** [notepad-plus-plus/notepad-plus-plus#9161]
If you scroll to the point where 1000th line is entering the view from the bottom, and this change would cause the horizontal scrollbar to appear due to a long line that was just touching the right size, then the screen starts flickering like crazy.

The reason why this happens is because when the 1000th line enters the screen, the left side to display the lines is no longer sufficient so it is widened. While doing so, the line that was "just fitting" will no longer fit, causing the horizontal scrollbar to appear. But once the horizontal scrollbar appears, the 1000th line is no longer visible, triggering a reduce on rows number left panel. Now, the horizontal scrollbar is no longer needed, so it disappears, causing the loop to start all over again.

To reproduce the issue, create a blank file (no need to save), make sure the width of the window is "just right" so that one more character would make the horizontal scrollbar appear. Then create more than 1000 empty lines and position your vertical scrolling so that the 1000th line is exactly the last one displayed.

**Comment by sasumner** [comment]
Another manifestation of #5670

**Comment by thg2k** [comment]
Not quite, it is about the same rows number panel but a whole different issue. Please try to reproduce, the screen starts literally vibrating. This is a bug not a feature request.

**Comment by sasumner** [comment]
>Not quite

Hmmm, it all seems predicated on :

>when the 1000th line enters the screen, the left panel that displays the line numbers is no longer sufficient, thus it is widened

I think if #5670 was fixed such that the margin doesn't widen/shrink, you wouldn't get into the situation you describe that comes after.

**Comment by thg2k** [comment]
> I think if #5670 was fixed such that the margin doesn't widen/shrink, you wouldn't get into the situation you describe that comes after.

#5670 is a feature request, and yes, if it is implemented it would clearly hide or soften the bug, depending on what would happen if you reach the maximum fitting number. Maybe the bug would reappear at line 10'000, or 100'000.

So again, this is a bug, and the fix is to break the resize/show scrollbar/resize again loop, for example by triggering the resizing of the line numbers pane only after 2 lines left the view instead of 1. That is, now when you are viewing line 1'000 and scroll up one line, so the last at the bottom of the view is 999, the view is resized, causing the horizontal scrollbar to disappear and the loop to go on. If this resize happens when the last line at the bottom is 998 the problem is fixed.

I still think it's unrelated to #5670.

**Comment by sasumner** [comment]
Okay, it can be unrelated to 5670 if you like. :-)

I don't see 5670 as a feature request (even though it says that). I consider a dynamically sizing margin to be a "bad thing" and thus my idea for 5670 was to eliminate the dynamic sizing entirely (which would preclude your bug from occurring).  But my plan was not accepted.

So your bug remains to be dealt with some other way.
But probably not by me. :-)

**Comment by MikeG621** [comment]
Ran into this bug myself for the first time today, my line 989 happens to be right at the window width, so the margin shift at 1000 causes the loop to start.

**Comment by sasumner** [comment]
@MikeG621 Try the constant-width setting for the line number margin?:

![image](https://user-images.githubusercontent.com/30118311/115158331-978cfe80-a05b-11eb-8be2-dbd9abc27eef.png)

**Comment by MikeG621** [comment]
That works, missed that setting addition.  Did verify that the dynamic loop
occurs at line 10,000 as well, so it's likely a general case.
--

**Comment by sasumner** [comment]
This should come above the following:

>That works, missed that setting addition.  Did verify that the dynamic loop
occurs at line 10,000 as well, so it's likely a general case.

Try the constant-width setting for the line number margin?

![image](https://user-images.githubusercontent.com/30118311/115162466-a8486f00-a071-11eb-9c55-b63a7f37d26c.png)

**Comment by rdipardo** [comment]
*F.Y.I.* There's a feature proposal on Scintilla's [mailing list](https://groups.google.com/g/scintilla-interest) that could mitigate this issue in the future:

> Applications may have a line number margin that widens or narrows depending on the number of digits needed to display all line numbers. A SC_UPDATE_LINE_COUNT flag for SCN_UPDATEUI can improve performance for this feature.
>
> <https://groups.google.com/g/scintilla-interest/c/O4R6zOtmO_M>

**Comment by nyamatongwe** [comment]
`SCN_UPDATEUI` may change the effects and performance a little but won't fix this problem. A choice of line number width needs to be made for the vertical scroll position and stuck to even after the minor change of horizontal scrollbar on/off. Its better to err on the wider line number side as a truncated line number will be a visible error. One possibility is to not decrease the line number width if the end display line number is just before the next power of 10.

That is, `if digits(lastLine) < digits(lastLine+1)` and `digits(lastLine)` would decrease width, use `digits(lastLine+1)`. This might be implemented by retaining the number of digits in a field or retrieving the width through `SCI_GETMARGINWIDTHN`.

The horizontal scrollbar height may be taller than a single line of text and that could lead to preferring `digits(lastLine+2)` or similar but its likely the above will improve behaviour enough.
